### PR TITLE
Jt/visualize attacks

### PR DESCRIPTION
**Pull request by RcColes**
The wait is finally over - we can actually see where we are attacking.

Also, this need a quite large rewrite of the stage rendering code. 

* We can now easily specify the two colors to use for the level by supplying them as arguments to `SystemStageRenderer.colorStage()`.
* The code for animating the world has been moved into a seperate function - `SystemStageRenderer.AnimateTiles()`
* The wall falling animation has been changed slightly, to be constant speed. This was done to clean up the 4 separate lists used in the previous implementation. It currently takes 1.5 seconds, but this can be adjusted on line `171` of `SystemStageRenderer.cpp`
* We can now easily recolour any tile of the floor by calling `SystemStageRenderer.setTileColor()` on it, with an `irr::video::SColor` and and `glm::ivec2` to specify the color and the position. This has no real performance cost, and subsequent calls will overwrite previous ones.
* ~~Performance has been increased slightly by tightening the clipping boundary for the world from `24, 7, 33` to `24, 7, 24`. I have tested this, and it does not appear to ever clip things that should be onscreen, although it is hard to check fully without a large open area on the map.~~
* Attacks are now rendered with a blue area, which briefly shows where the attack has affected.
* Delayed attacks are now rendered with a blue area, becoming darker as the attack is imminent. This appears to work well, and should help with players telling what is going on.

Unintended changes:

* The floor is now red and purple, because those were already defined in main. Another thing for the colour rework.

TODO / future improvemnts:

* The attack colours need changing, but I'd like to do this as part of a full overhaul of the colour system, which will go in another pull request. For now, the attacks are visually distinct, which should be enough for the demo.

### CODE REVIEW

File: src/client/include/client/misc/Color.hpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
+#ifndef CLIENT_MISC_COLOR
+#define CLIENT_MISC_COLOR
```
[the remark this answers is not in the document]
Comment (jnterry) [reply src/client/include/client/misc/Color.hpp]: Code should be consistent, and the convention is the american spelling. Happy times :)

[the remark this answers is not in the document]
Comment (AnthonyWharton) [reply src/client/include/client/misc/Color.hpp]: It's a great way to make me sad

[the remark this answers is not in the document]
Comment (grant-m-s) [reply src/client/include/client/misc/Color.hpp]: Ding, Ding, Round Two... Fight!

File: src/client/include/client/system/SystemStageRenderer.hpp [hunk 2]
```diff
@@ -13,65 +13,73 @@
...
-	                 int                        j,
-	                 glm::ivec2                 playerpos,
-                   irr::video::SColor colour);
+
+	void Update(irr::scene::ISceneManager *smgr,
+                    irr::video::IVideoDriver * driver,
+                    glm::ivec2                 playerpos);
+
+	void colorStage(int                        j,
+                        irr::video::SColor         C1,
```
Comment (AnthonyWharton) [inline src/client/include/client/system/SystemStageRenderer.hpp]: If we're going to be picky about whitespace.. here is a place to be picky.

File: src/client/include/client/system/SystemStageRenderer.hpp [hunk 3]
```diff
@@ -13,65 +13,73 @@
...
 	inline void addFloorTilesToScene(irr::scene::ISceneManager *smgr,
 	                                 irr::video::IVideoDriver * driver,
-	                                 tempo::ComponentStage &    stage,
-																 glm::ivec4 colour1, glm::ivec4 colour2);
-
-
-	inline bool checkerBoardPattern(irr::video::IVideoDriver *driver,
-	                                int                       i,
-	                                int                       j);
+	                                 tempo::ComponentStage &    stage);
```
Comment (AnthonyWharton) [inline src/client/include/client/system/SystemStageRenderer.hpp]: More dodgy spacing or is that the github preview?

File: src/client/src/misc/Color.cpp [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#include <client/misc/Color.hpp>
+
+namespace client {
+
+	void init_palettes()
```
Comment (AnthonyWharton) [inline src/client/src/misc/Color.cpp]: Does this have a purpose?

File: src/client/src/system/SystemAttack.cpp [hunk 5]
```diff
@@ -33,6 +34,8 @@ void SystemAttack::processServerResponses(anax::World &w)
 		if (id.isNull()) continue;
 		anax::Entity e(w, id);
 
+		printf("Processing server attack message\n");
```
Comment (AnthonyWharton) [inline src/client/src/system/SystemAttack.cpp]: Please remove extra prints before merging to master.

**Comment by grant-m-s** [comment]
Yeah it seems to attack in the direction the enemy is facing, so because they start facing north it colours north, the moment it turns it lights up correct

File: src/client/src/system/SystemStageRenderer.cpp [hunk 6]
```diff
@@ -78,153 +75,168 @@ void SystemStageRenderer::setup(irr::scene::ISceneManager *smgr, irr::video::IVi
...
-	bool firstRun = false;
-
-	if (this->currentHeight.size() <= 1) {
-		this->currentHeight = std::vector<float>(this->tile_nodes.size());
-		firstRun            = true;
-	}
-
+void SystemStageRenderer::colorStage(int                        step,
+                                     irr::video::SColor         C1,
+				     irr::video::SColor         C2)
```
Comment (AnthonyWharton) [inline src/client/src/system/SystemStageRenderer.cpp]: Spacing thing again I guess :man_shrugging:

**Comment by AnthonyWharton** [comment]
@Codrin-Popa The FPS is exactly the same for me. Were you running on your laptop when not plugged to power?

File: src/client/include/client/misc/Color.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (RcColes) [reply src/client/include/client/misc/Color.hpp]: American spelling is standard in CS, I'm with Jamie

**Comment by RcColes** [comment]
AI should be synced properly now - it was a bug with how their rotation code interacted with the netcode

**Comment by Codrin-Popa** [comment]
@AnthonyWharton FPS is 21 with the laptop plugged in

**Comment by RcColes** [comment]
I'm not getting and FPS changes either, could @grant-m-s and @jnterry weight in on their experience?

**Comment by grant-m-s** [comment]
on linux im still getting 60, ill give it a run in the lab tomorrow morning and give you the important verdict

**Comment by jnterry** [comment]
I'm also still getting 60fps. Attack direction is now correct so I'm happy

File: src/client/include/client/system/SystemStageRenderer.hpp (hunk 3 above)
Comment (RcColes) [reply src/client/include/client/system/SystemStageRenderer.hpp]: Think you're confused here - remember the red lines are being removed

File: src/client/src/misc/Color.cpp (hunk 4 above)
Comment (RcColes) [reply src/client/src/misc/Color.cpp]: This is for the colour redesign - strictly speaking it shouldn't really be in this PR but it's basically dead code so won't actually affect anything.

**Comment by AnthonyWharton** [comment]
Just merging in master now manually to sort conflicts, Can I ask how this is going to affect the HUD/colours for tomorrow?

**Comment by AnthonyWharton** [comment]
Should be good to go, but the previous comment/question stands before I make a review.

**Comment by RcColes** [comment]
Currently the only change is to the floor tiles - if you want to adjust this you can pass arguments into the function I mentioned in the description.

**Comment by RcColes** [comment]
Probably easier to just patch this one so that the floor is colored like it used to be - should be easy but I'm not sure how it used to work so @AnthonyWharton or @Codrin-Popa might be able to do it more easily

**Comment by Codrin-Popa** [comment]
@AnthonyWharton  Floor is now coloured as it used to be

**Pull request merged**